### PR TITLE
[netcore] Implement Array.GetElementSize as intrinsic

### PR DESCRIPTION
**Pull request by filipnavara**

### CODE REVIEW

File: mono/mini/intrinsics.c [hunk 1]
```diff
@@ -755,6 +756,13 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
 			EMIT_NEW_BIALU_IMM (cfg, ins, OP_PADD_IMM, dreg, args [0]->dreg, MONO_STRUCT_OFFSET (MonoArray, vector));
 			return ins;
 		}
+		else if (!strcmp (cmethod->name, "GetElementSize")) {
+			int dreg = alloc_ireg (cfg);
+			int vt_reg = alloc_preg (cfg);
+			MONO_EMIT_NEW_LOAD_MEMBASE_FAULT (cfg, vt_reg, args [0]->dreg, MONO_STRUCT_OFFSET (MonoObject, vtable));
+			EMIT_NEW_LOAD_MEMBASE (cfg, ins, OP_LOAD_MEMBASE, dreg, vt_reg, MONO_STRUCT_OFFSET (MonoVTable, klass));
+			return mono_emit_jit_icall (cfg, mono_array_element_size, &ins);
```
Comment (EgorBo) [inline mono/mini/intrinsics.c]: can't you just access size directly without the icall?

Comment (filipnavara) [reply mono/mini/intrinsics.c]: That was my first thought but it may not be initialized. Not sure if there's an easy solution for that.

Comment (EgorBo) [reply mono/mini/intrinsics.c]: Well, you can emit a condition then 🙂, and emit the icall if `m_class_is_size_inited` is 0 as a slow path.

Comment (vargaz) [reply mono/mini/intrinsics.c]: It should be initialized already if there are instances of the array.

Comment (filipnavara) [reply mono/mini/intrinsics.c]: I'll rework it.

Comment (vargaz) [reply mono/mini/intrinsics.c]: The icall had to initialize it because it was working on Type objects.

**Comment by filipnavara** [comment]
@monojenkins build Linux x64 Checked Private Types Build

File: mono/mini/intrinsics.c [hunk 2]
```diff
@@ -755,6 +756,15 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
 			EMIT_NEW_BIALU_IMM (cfg, ins, OP_PADD_IMM, dreg, args [0]->dreg, MONO_STRUCT_OFFSET (MonoArray, vector));
 			return ins;
 		}
+		else if (!strcmp (cmethod->name, "GetElementSize")) {
+			int vt_reg = alloc_preg (cfg);
+			int class_reg = alloc_preg (cfg);
+			int sizes_reg = alloc_preg (cfg);
```
Comment (EgorBo) [inline mono/mini/intrinsics.c]: `sizes_reg` needs to be `alloc_ireg`? also I guess you can re-use `vt_reg` for the second load (`vt_reg, vt_reg`).

Comment (filipnavara) [reply mono/mini/intrinsics.c]: You're right. It ends up implemented using the same code but I should be using the right one. Reusing the register would probably work but not sure how beneficial it is. For LLVM we convert to SSA anyway.

Comment (EgorBo) [reply mono/mini/intrinsics.c]: yeah I know, it's just a nit - less code

**Comment by filipnavara** [comment]
Hmm, I forgot I need to implement it in the interpreter too.

**Comment by filipnavara** [comment]
@monojenkins build Linux x64 Checked Private Types Build

File: mono/mini/interp/interp.c [hunk 3]
```diff
@@ -5338,6 +5338,13 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
 			ip++;
 			MINT_IN_BREAK;
 		}
+		MINT_IN_CASE(MINT_ARRAY_ELEMENT_SIZE) {
+			MonoObject* const o = sp [-1].data.o;
+			NULL_CHECK (o);
+			sp [-1].data.i = mono_class_array_element_size (mono_object_class (sp [-1].data.p));
+			ip++;
```
Comment (vargaz) [inline mono/mini/interp/interp.c]: This should use 'o'.

**Comment by vargaz** [comment]
The failure is relevant:
intrinsics.c:765:4: error: invalid use of incomplete typedef 'MonoClass {aka struct _MonoClass}'
    EMIT_NEW_LOAD_MEMBASE (cfg, ins, OP_LOADI4_MEMBASE, sizes_reg, class_reg, MONO_STRUCT_OFFSET (MonoClass, sizes));

**Comment by filipnavara** [comment]
Thanks, I missed that in the log file. Not sure how to solve it though. (/cc @lambdageek)

**Comment by jaykrell** [comment]
Just missing an include?

**Comment by jaykrell** [comment]
Probably over in class-accessors.c, you'd do something like
```
// Use this power responsibly.
extern const int mono_private_offsetof_MonoClass_sizes = offsetof(MonoClassDef, sizes);
```
or more like what you have, i.e. for cross builds.

**Comment by vargaz** [comment]
Use m_class_offsetof_sizes instead of MONO_STRUCT_OFFSET.

**Comment by filipnavara** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by filipnavara** [comment]
@monojenkins build Linux x64 Interpreter

**Pull request merged**